text/0031-integ-testing.md

```diff
@@ -0,0 +1,300 @@
+# Integration testing for CDK apps
+
+* **Original Author(s):**: [@nija-at](https://github.com/nija-at)
+* **Tracking Issue**: [#31](https://github.com/aws/aws-cdk-rfcs/issues/31)
+* **API Bar Raiser**: @{BAR_RAISER_USER}
+
+CDK app developers and construct library authors can write integration
+tests that can be run locally or in their CI/CD pipeline.
+
+## Working Backwards - README.md
+
+### contest
+
+🌩️ contest tests can verify that your CDK app can be deployed successfully on AWS with the desired resources.
+
+🤖 contest tests can verify that your CDK app functions as expected, such as assertions against a public API endpoint.
+
+🕰️ contest tests can verify that your CDK app can be updated from a previous version of the app.
+
+🧑‍💻 Write contest tests in any CDK supported language, and execute them as part of your CI pipeline.
+
+📸 Capture snapshots of your CDK apps to track how they are impacted by changes.
+
+### Writing Tests
+
+contest tests can be written in the same programming language as your CDK project.
+
+Let's start with the below CDK construct that creates an artifact repository in an S3 bucket,
+and notifies of new artifacts to an SQS queue.
+
+```ts
+// repo.ts
+import { Construct } from 'constructs';
+import { Bucket, EventType, IBucket } from '@aws-cdk/aws-s3';
+import { IQueue, Queue } from '@aws-cdk/aws-sqs';
+import { SqsDestination } from '@aws-cdk/aws-s3-notifications';
+
+export class ArtifactRepo extends Construct {
+  public readonly artifacts: IBucket;
+  public readonly publishNotifs: IQueue;
+  
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    this.artifacts = new Bucket(this, 'Artifacts');
+    this.publishNotifs = new Queue(this, 'PublishNotifs');
+
+    this.artifacts.addEventNotification(EventType.OBJECT_CREATED_PUT,
+      new SqsDestination(this.publishNotifs))
+  }
+}
+```
+
+The following integ test invokes puts an object in the S3 bucket and verifies the message in the SQS queue.
+
+```ts
+// repo.contest.ts
+import { ArtifactRepo } from './records';
+import { Test } from '@aws-cdk/contest';
+
+const repo = new ArtifactRepo(app, 'RequestRecord');
+
+const repoTest = Test.forConstruct(repo);
+
+repoTest.assertAws.
+  s3.putObject({ bucket: stack.artifacts, key: '...', body: '...' });
```

Comment:
> This feels like test setup, which makes the "assert" part of this statement read a bit awkwardly.

Hmm yes, you make a good point. Maybe just dropping the "assert" phrasing in the API will make this cleaner. I'll take a stab at this in the next rev.

> this may lead users to attempt to modify the return values of the SDK calls in the test file
> I think this will be the most confusing part of the paradigm you are describing here.

Good observation. I came to the similar realization after I published this. There's further work needed here.